string.c

```diff
@@ -9159,6 +9159,73 @@ rb_str_end_with(int argc, VALUE *argv, VALUE str)
     return Qfalse;
 }
 
+/*
+ *  call-seq:
+ *     str.delete_prefix!(substr)   -> self or nil
+ *
+ *  Deletes leading <code>substr</code> from <i>str</i>, returning
+ *  <code>nil</code> if no change was made.
+ *
+ *     "hello".delete_prefix!("hel")  #=> lo
+ *     "hello".delete_prefix!("llo")  #=> nil
+ */
+
+static VALUE
+rb_str_delete_prefix_bang(VALUE str, VALUE substr)
+{
+    long loffset;
+
+    str_modify_keep_cr(str);
+    if (rb_str_start_with(1, &substr, str) == Qtrue) {
+        loffset = RSTRING_LEN(substr);
+    }
+    else {
+        return Qnil;
+    }
+    if (loffset > 0) {
+        char *start, *s;
+        rb_encoding *enc;
+        long olen, len;
+
+        RSTRING_GETMEM(str, start, olen);
+        len = olen - loffset;
+        s = start + loffset;
+        memmove(start, s, len);
+        STR_SET_LEN(str, len);
+#if !SHARABLE_MIDDLE_SUBSTRING
+        enc = STR_ENC_GET(str);
+        TERM_FILL(start + len, rb_enc_mbminlen(enc));
+#endif
+        return str;
```

Comment:
```
$ ruby -e's="あ"; p s.valid_encoding?; s.chomp!("\x82"); p s; p s.valid_encoding?'
true
"あ"
true
```

Because `String#chomp!` behaves as above, I fixed `delete_prefix!` to work as:

```
$ ruby -e's="あ"; p s.valid_encoding?; s.delete_prefix!("\xe3"); p s; p s.valid_encoding?'
true
"あ"
true
```